smarts/core/lanepoints.py

```diff
@@ -0,0 +1,564 @@
+from smarts.core.utils.math import (
+    lerp,
+    radians_to_vec,
+    signed_dist_to_line,
+    squared_dist,
+    vec_2d,
+    vec_to_radians,
+)
+
+
+@dataclass(frozen=True)
+class LanePoint:
+    """a.k.a. MapPoint:  static, based only on a map, always in a Lane"""
+
+    pos: np.ndarray  # Point positioned on center of lane
+    heading: Heading  # Heading angle of lane at this point (radians)
+    lane_width: float  # Width of lane at this point (meters)
+    speed_limit: float  # Lane speed in m/s
+    lane_id: str  # ID of lane under lanepoint
+    lane_index: int  # Index of the lane this lanepoint is over. 0 is the outer(right) most lane.
+
+    def dist_to(self, p) -> float:
+        """Calculates straight line distance to the given 2D point"""
+        return np.linalg.norm(self.pos - vec_2d(p))
+
+    def relative_heading(self, h: Heading) -> Heading:
+        """Computes relative heading between the given angle and the lanepoint heading
+
+        Returns:
+            relative_heading: [-pi..pi]
+        """
+        assert isinstance(
+            h, Heading
+        ), "Heading h ({}) must be an instance of smarts.core.coordinates.Heading".format(
+            type(h)
+        )
+
+        return self.heading.relative_to(h)
+
+    def signed_lateral_error(self, p) -> float:
+        """Returns the signed lateral distance from the given point to the
+        line formed by the lanepoint position and the lanepoints heading.
+
+        Negative signals right of line and Positive left of line
+        """
+        return signed_dist_to_line(p, self.pos, self.heading.direction_vector())
+
+    @property
+    def pose(self) -> Pose:
+        return Pose.from_center(tuple(self.pos), self.heading)
+
+    def __hash__(self):
+        return hash(
+            (
+                *self.pos,
+                self.heading,
+                self.lane_width,
+                self.speed_limit,
+                self.lane_id,
+                self.lane_index,
+            )
+        )
+
+    def __eq__(self, other):
+        if not isinstance(other, LanePoint):
+            return False
+        return (
+            (self.pos == other.pos).all()
+            and self.heading == other.heading
+            and self.lane_width == other.lane_width
+            and self.speed_limit == other.speed_limit
+            and self.lane_id == other.lane_id
+            and self.lane_index == other.lane_index
+        )
+
+
+class LinkedLanePoint(NamedTuple):
+    lp: LanePoint = None
+    is_shape_lp: bool = False
+    nexts: List[LinkedLanePoint] = []  # list of next immediate LanePoint(s)
+    # it's a list of LanePoints because a path may branch at junctions
+
+    def __hash__(self):
+        ## distinguish between different continuations here too
+        ## so the lru_cache on _lanepoint_paths_starting_at_lanepoint() below
+        ## doesn't return the wrong set of LanePoints.
+        return hash(self.lp) + sum(hash(nlp.lp) for nlp in self.nexts)
```

Comment:
The sum of uniform random variables is not uniform (it's triangular) so you'll get some nasty hash collisions perf issues if this hash is used heavily.

It'd be safer to hash the tuple of the lp's, something like:
```suggestion
        return hash((self.lp, tuple(nlp.lp for nlp in self.nexts))
```

Reply:
@davidrusu Hi David!  Good to see you here!  :)
And thanks for the review.  Yep, you're right.  Will make the change now.

Reply:
Likewise! good to see you involved with SMARTS!